### ISSUE TITLE
`unnecessary_to_owned` suggestion causes error when iterator is re-borrowed mutably within a loop

### ISSUE THREAD

**Issue by teor2345** [rust-lang/rust-clippy#8148]
### Summary

When running:
```sh
cargo clippy --fix
```

On https://github.com/ZcashFoundation/zebra/tree/410133435e875336875aab32c338a284470ceab4/zebra-chain

I get an error:
> after fixes were automatically applied the compiler reported errors

### Lint Name

unnecessary_to_owned

### Reproducer

I tried this code:

```rust
for mut input in transaction.inputs().to_vec().into_iter() { 
    // use input and &mut transaction
}
```

https://github.com/ZcashFoundation/zebra/blob/410133435e875336875aab32c338a284470ceab4/zebra-chain/src/block/arbitrary.rs#L554-L578

I saw this happen:

```
after fixes were automatically applied the compiler reported errors within these files:

  * zebra-chain/src/block/arbitrary.rs
...
The following errors were reported:                                                                                                                                                            
error[E0502]: cannot borrow `transaction` as mutable because it is also borrowed as immutable                                                                                                  
   --> zebra-chain/src/block/arbitrary.rs:559:17                                                                                                                                               
    |                                                                                                                                                                                          
554 |     for mut input in transaction.inputs().iter().cloned() {                                                                                                                              
    |                      ------------------------------------                                                                                                                                
    |                      |                                                                                                                                                                   
    |                      immutable borrow occurs here                                                                                                                                        
    |                      immutable borrow later used here                                                                                                                                    
...                                                                                                                                                                                            
559 |                 &mut transaction,                                                                                                                                                        
    |                 ^^^^^^^^^^^^^^^^ mutable borrow occurs here   
```

I expected to see this happen:

Either no lint, or a fix that compiles correctly.

### Version

```text
rustc 1.59.0-nightly (7abab1efb 2021-12-17)
binary: rustc
commit-hash: 7abab1efb21617ba6845fa86328dffa16cfcf1dc
commit-date: 2021-12-17
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by smoelius** [comment]
I'll get started on a fix right away.

**Comment by teor2345** [comment]
Thanks!

I'm not sure if this will help, but here's the workaround we ended up merging:
```rust
    let original_inputs = transaction.inputs().to_vec();
    for mut input in original_inputs.into_iter() {
        // use input and &mut transaction
    }
```

We felt it was clearer than the original code.

**Comment by smoelius** [comment]
That is helpful. Thank you, @teor2345.

**Comment by rhysd** [comment]
I encountered this too. Here is another smaller reproduction only with prelude:

```rust
fn repro() {
    let mut v = vec![1];
    v.to_vec().into_iter().map(|i| v.push(i));
}
```
